[doc][core][cgraph] Complete Compiled Graph docs (#51206)
ray-project/ray: doc/source/ray-core/compiled-graph/troubleshooting.rst

This page contains common issues and solutions for Compiled Graph execution.

Limitations
-----------

> Compiled Graph is a new feature and has some limitations. Different from classic Ray Core:

Reviewer: ```suggestion
Compiled Graph is a new feature and has some limitations:
```

- Invoking Compiled Graph

  - Only the process that compiles the Compiled Graph may call it.

>   - A Compiled Graph has a maximum number of in-flight executions. Normally when using classic Ray
>     Core, if there aren't enough resources at the time of ``dag.execute()``, Ray will queue the

Reviewer: ```suggestion
  - A Compiled Graph has a maximum number of in-flight executions. When using the DAG API, 
     if there aren't enough resources at the time of ``dag.execute()``, Ray will queue the
```

>     tasks for later execution. Ray Compiled Graph currently does not support such queuing past their
>     maximum capacity. Therefore, you may need to consume some results using ``ray.get()``

Reviewer: ```suggestion
    tasks for later execution. Ray Compiled Graph currently doesn't support queuing past its
    maximum capacity. Therefore, you may need to consume some results using ``ray.get()``
```
I think a link to some what this maximum capacity is could also be useful

    before submitting more executions. As a stopgap,
    ``dag.execute()`` throws a ``RayCgraphCapacityExceeded`` exception if the call takes too long.

>     In the future, we plan to introduce better error handling and queuing for Compiled Graph.

Reviewer: ```suggestion
    In the future, Compiled Graph may have better error handling and queuing.
```

- Compiled Graph Execution

  - Ideally, you should try not to execute other tasks on the actor while it is participating in a Compiled Graph.
    Compiled Graph tasks will be executed on a **background thread**. Any concurrent tasks
    submitted to the actor can still execute on the main thread, but you are responsible for
    synchronization with the Compiled Graph background thread.

  - For now, actors can only execute one Compiled Graph at a time. To execute a different Compiled Graph

>     on the same actor, the current Compiled Graph must be torn down.

Reviewer: ```suggestion
    on the same actor, you must teardown the current Compiled Graph.
```

    See :ref:`Return NumPy arrays <troubleshoot-numpy>` for more details.

- Passing and getting Compiled Graph results (:class:`CompiledDAGRef <ray.experimental.compiled_dag_ref.CompiledDAGRef>`)
  
  - Compiled Graph results can't be passed to another task or actor. This restriction may be loosened
    in the future, but for now, it allows for better performance because the backend knows
    exactly where to push the results.

  - ``ray.get()`` can be called at most once on a :class:`CompiledDAGRef <ray.experimental.compiled_dag_ref.CompiledDAGRef>`. An exception will be raised if
    it is called twice on the same :class:`CompiledDAGRef <ray.experimental.compiled_dag_ref.CompiledDAGRef>`. This is because the underlying memory for

>     the result may need to be reused for a future DAG execution; restricting ``ray.get()`` to once

Reviewer: ```suggestion
    the result may need to be reused for a future DAG execution. Restricting ``ray.get()`` to once
```

    per reference simplifies the tracking of the memory buffers.

  - If the value returned by ``ray.get()`` is zero-copy deserialized, then subsequent executions
    of the same DAG will block until the value goes out of scope in Python. Thus, if you hold onto

>     zero-copy deserialized values returned by ``ray.get()``, and you try to execute the DAG above

Reviewer: ```suggestion
    zero-copy deserialized values returned by ``ray.get()``, and you try to execute the Compiled Graph above
```

>     its max concurrency, you may deadlock. This case will be detected in the future, but for now

Reviewer: ```suggestion
    its max concurrency, it may deadlock. This case will be detected in the future, but for now
```

    you will receive a ``RayChannelTimeoutError``.
    See :ref:`Explicitly teardown before reusing the same actors <troubleshoot-teardown>`
    for more details.

- Collective operations

>   - Compiled Graph supports all-reduce collective operation now, but not yet other collective operations.

Reviewer: ```suggestion
  - Compiled Graph only supports the all-reduce collective operation for now.
```
